sw/device/lib/runtime/print.c

```diff
@@ -68,6 +71,151 @@ void base_set_stdout(buffer_sink_t out) {
+  }
+
+  // Add the magic bytes.
+  for (size_t i = 0; i < 4; ++i) {
+    frame_bytes[i] = (kSpiDeviceFrameMagicNumber >> (i * 8)) & 0xff;
+  }
+  // Add the frame number.
+  for (size_t i = 0; i < 4; ++i) {
+    frame_bytes[i + 4] = (spi_device_frame_num >> (i * 8)) & 0xff;
+  }
+  // Add the data length.
+  for (size_t i = 0; i < 4; ++i) {
+    frame_bytes[i + 8] = (len >> (i * 8)) & 0xff;
+  }
+
+  // Construct the frame data packet.
+  // Add the data and pad bytes.
+  for (size_t i = 0; i < ((len + 3u) & ~3u); ++i) {
+    if (i < len) {
+      frame_bytes[i + 12] = buf[i];
+    } else {
+      frame_bytes[i + 12] = 0xff;
+    }
+  }
+
+  uint32_t available_buffer_size = 0;
+  do {
+    uint32_t last_read_address = 0;
+    if (dif_spi_device_get_last_read_address(spi_device, &last_read_address) !=
+        kDifOk) {
+      return 0;
+    }
+
+    // Adjust the last read address. The host continuously reads from the read
+    // buffer, unaware of whether a new frame has arrived. This could result in
+    // the reported last_read_address being the header size ahead of the actual
+    // address of the last valid frame if all the frames in the read buffer has
+    // been consumed by the host. While it's harmless to adjust the last read
+    // address even if the reported value is correct, doing so might temporarily
+    // underestimate the available buffer size by the size of a header.
+    uint32_t adjusted_last_read_address =
+        (kSpiDeviceReadBufferSizeBytes + last_read_address -
+         kSpiDeviceFrameHeaderSizeBytes) %
+        kSpiDeviceReadBufferSizeBytes;
+    uint32_t next_read_address = ((adjusted_last_read_address + 1) & ~3u) %
+                                 kSpiDeviceReadBufferSizeBytes;
+
+    if (next_read_address > next_write_address) {
+      available_buffer_size = next_read_address - next_write_address - 1;
+    } else {
+      available_buffer_size =
+          next_read_address +
+          (kSpiDeviceReadBufferSizeBytes - next_write_address) - 1;
+    }
+  } while (kFrameSizeBytes > available_buffer_size);
+
+  // Write frame (handle wrap-around)
+  if (next_write_address + kFrameSizeBytes > kSpiDeviceReadBufferSizeBytes) {
+    size_t first_part_size = kSpiDeviceReadBufferSizeBytes - next_write_address;
+    size_t second_part_size = kFrameSizeBytes - first_part_size;
+    if (dif_spi_device_write_flash_buffer(
+            spi_device, kDifSpiDeviceFlashBufferTypeEFlash, next_write_address,
+            first_part_size, frame_bytes) != kDifOk) {
+      return 0;
+    }
+    if (dif_spi_device_write_flash_buffer(
+            spi_device, kDifSpiDeviceFlashBufferTypeEFlash, 0, second_part_size,
+            (uint8_t *)(frame_bytes + first_part_size)) != kDifOk) {
+      return 0;
+    }
+  } else {
+    if (dif_spi_device_write_flash_buffer(
+            spi_device, kDifSpiDeviceFlashBufferTypeEFlash, next_write_address,
+            kFrameSizeBytes, frame_bytes) != kDifOk) {
+      return 0;
+    }
+  }
```

Comment:
This can be slightly simplified to improve readability.
```suggestion
   size_t round_size = next_write_address + kFrameSizeBytes;
   round_size = round_size > kSpiDeviceReadBufferSizeBytes ? kSpiDeviceReadBufferSizeBytes
                                                           : round_size;
   TRY(dif_spi_device_write_flash_buffer(
           spi_device, kDifSpiDeviceFlashBufferTypeEFlash, next_write_address,
           round_size, frame_bytes) != kDifOk) );
   // Handle wrap-around.
   if ( round_size < kFrameSizeBytes){
     TRY(dif_spi_device_write_flash_buffer( spi_device, kDifSpiDeviceFlashBufferTypeEFlash,
           0, kFrameSizeBytes - round_size,
           (uint8_t *)(frame_bytes + round_size)) != kDifOk))
   }
  ```

Reply:
Thanks for the suggestion! If I understand the idea correctly, an additional line `round_size-=next_write_address;` needs to be added between line 215 and 216 in the suggestion. The implementation has been modified to align with the proposed suggestion, along with some additional modifications. Let me know if the new implementations meet your expectations.

Reply:
I would like to add a note that the “TRY” macro wasn’t employed here because the function returns a `size_t` value.

Reply:
Yes, let's avoid the use of TRY in this function as that is not the return type and this is a runtime lib. @engdoreis